### ISSUE TITLE
tf.dynamic_partition cannot have dynamic num_partitions

### ISSUE THREAD

**Issue by Chasearmer** [tensorflow/tensorflow#19957]
I am attempting to use `tf.dynamic_partition` as so:
`tf.dynamic_partition(features, membership, tf.reduce_max(membership))`
Unfortunately, however, it is throwing this error
>TypeError: Expected int for argument 'num_partitions' not <tf.Tensor 'Max:0' shape=() dtype=int32>.

Can `tf.dynamic_partition` not receive a dynamic input for num_partitions?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by Chasearmer** [comment]
Have I written custom code: Yes
OS Platform and Distribution: Linux Ubuntu 16.04
TensorFlow installed from: Conda
TensorFlow version: 1.5.0
Bazel version: NA
CUDA/cuDNN version: NA
GPU model and memory: NA
Exact command to reproduce: Written above

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cBournhonesque** [comment]
I have the same issues. I thought we could use a Tensor instead of an integer anywhere in the code?
It would be nice to have a dynamic number of partitions.

**Comment by aselle** [comment]
@skyewm, are you familiar with functional ops?

**Comment by anjabeth** [comment]
I'm having the same problem as well - the documentation says "partitions: A Tensor of type int32. Any shape. Indices in the range [0, num_partitions)", but it doesn't seem to accept a Tensor.

**Comment by aselle** [comment]
No, num_partitions is an attribute which means it cannot be driven by a tensor.  See the op_def below:
```op {
  name: "DynamicPartition"
  input_arg {
    name: "data"
    type_attr: "T"
  }
  input_arg {
    name: "partitions"
    type: DT_INT32
  }
  output_arg {
    name: "outputs"
    type_attr: "T"
    number_attr: "num_partitions"
  }
  attr {
    name: "num_partitions"
    type: "int"
    has_minimum: true
    minimum: 1
  }
  attr {
    name: "T"
    type: "type"
  }
} 
```
If you can't work around it, you could certainly make a DynamicPartitionV2 that takes tensors and not attributes, but there is usually a cost to this in terms of shape inference. This is why there is a separation between the two.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!

**Comment by facaiy** [comment]
@aselle Could you please comment?

**Comment by aselle** [comment]
is N constant for the graph?

**Comment by ybsave** [comment]
@aselle Thank you for your response. N is not constant. Actually, N is my dynamic input of batch_size. However, due to the non-divisible of the total number, the last batch will have a number of samples not equal to "batch_size". Also, I do not want to throw away the remaining few samples. Therefore, N is not constant.

**Comment by casiciaco** [comment]
Will this functionality be extended ?

**Comment by randolf-scholz** [comment]
I think this is only possible with eager execution at the moment

```python
import tensorflow as tf
tf.enable_eager_execution()

def dynamic_partition(inputs, labels):
    '''
    Splits tensor along first axis according to parition label
    
    Input: - inputs: tensor(shape=(N,...), dtype=T)
           - labels: tensor(shape=(N), dtype=?) with M<=N unique values
             the k-th unique value occurs exactly N_k times
    Returns: outputs: [tensor(shape=(N_1,...), dtype=T), ..., tensor(shape=(N_M,...), dtype=T)]
    '''
    classes, idxs = tf.unique(labels)
    
    indices = tf.range(tf.size(labels))
    
    classes = tf.unstack(classes)
    
    index_partition = [tf.boolean_mask(indices, tf.equal(labels, cl)) for cl in classes]

    input_partition = [tf.gather(x, cl_idx) for cl_idx in index_partition]
         
    return input_partition


x = tf.constant([[0,0], [1,1], [2,2], [3,3], [4, 4], [5,5], [6,6], [7,7], [8,8]], dtype=tf.float32)
labels = tf.constant(["A", "A", "B", "B", "B", "C", "A", "B", "C"])
_, idxs = tf.unique(labels)

print(dynamic_partition(x, labels))
print(tf.dynamic_partition(x, idxs, 3))
```

**Comment by sushreebarsa** [comment]
@Chasearmer Could you please refer to the[ link ](https://www.tensorflow.org/api_docs/python/tf/dynamic_partition),similar [ issue ](https://stackoverflow.com/questions/41408561/dynamic-partition-with-dynamic-num-partitions)& Please let us know if it helps? Thanks!
